### ISSUE TITLE
Reference count on closures

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3573]
Migrated from [rt.perl.org#6031](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6031) (status was 'resolved')

Searchable as RT6031$

**Comment by p5pRT** [comment]
### From me-01@ton.iguana.be



While writing a big application\, I was suffering from memory leaks\.
In the end\, I reduced it to this&#8203;:

\#\! /usr/bin/perl \-w
sub DESTROY \{ print STDERR "DESTROY\\n"\}
END \{ print STDERR "Global cleanup after this\\n"\}
sub fun \{
&nbsp;    my $canary =bless\[\]\, "main";
&nbsp;    return sub \{
&nbsp;        my $probe;
&nbsp;        \[sub \{$probe\}\]; \# This line
&nbsp;        die "Canary" unless $canary;
&nbsp;    \}
\}
fun\(\);
fun\(\);

The canary is used to see when the closure gets destroyed\.
In this form each call to fun will survive to global cleanup\.
If you however remove the line marked with "This line"\, the closures
will die as expected\, as soon as fun returns\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by ton at Wed Aug  9 23:24:07 CEST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.0-test5, archname=i686-linux
    uname='linux quasar 2.4.0-test5 #2 sat jul 29 04:45:14 cest 2000 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=undef 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -fomit-frame-pointer', gccversion=egcs-2.91.66 19990314 (egcs-1.1.2 release), gccosandvers=
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/opt/gnu/lib'
    libpth=/usr/local/lib /opt/gnu/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.6.0/i686-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib -L/opt/gnu/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i686-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/usr/local/jdk116_v5/bin:/home/oracle/product/8.1.5/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @floatingatoll

\#\! /usr/bin/perl \-w
sub DESTROY \{ print STDERR "DESTROY\\n"\}
END \{ print STDERR "Global cleanup after this\\n"\}
sub fun \{
&nbsp;    my $canary =bless\[\]\, "main";
&nbsp;    return sub \{
&nbsp;        my $probe;
&nbsp;        \[sub \{$probe\}\]; \# This line
&nbsp;        die "Canary" unless $canary;
&nbsp;    \}
\}
fun\(\);
fun\(\);
print "About to die\.\\n";
\_\_END\_\_

This feels like a circular reference issue; any gurus wish to confirm/deny that?  That single line would be where the circular ref
would be created\.

R\.

**Comment by p5pRT** [comment]
### From @iabyn

\(Just reviewing old bugs\)\.

This is fixed in bleedperl\, probably by patch \#18302

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
